lib/archive/untar.go

```diff
@@ -0,0 +1,135 @@
+// Copyright 2016-2017 VMware, Inc. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//    http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package archive
+
+import (
+	"archive/tar"
+	"fmt"
+	"io"
+	"os"
+	"path/filepath"
+	"strings"
+
+	"github.com/vmware/vic/pkg/trace"
+)
+
+const (
+	// fileWriteFlags is a collection of flags configuring our writes for general tar behavior
+	//
+	// O_CREATE = Create file if it does not exist
+	// O_TRUNC = truncate file to 0 length if it does exist(overwrite the file)
+	// O_WRONLY = We use this since we do not intend to read, we only need to write.
+	fileWriteFlags = os.O_CREATE | os.O_TRUNC | os.O_WRONLY
+)
+
+// Untar untars the given tarstream(if it is a tar stream) on the local filesystem based on the unpackPath in the path spec
+//
+// the pathSpec will include the following elements
+// - include : any tar entry that has a path below(after stripping) the include path will be written
+// - strip : The strip string will indicate the
+// - exlude : marks paths that are to be excluded from the write operation
+// - target : marks the the write path that will be tacked onto the "unpackPath". e.g /tmp/unpack + /my/target/path = /tmp/unpack/my/target/path
+func Untar(op trace.Operation, tarStream io.ReadCloser, filter *FilterSpec, unpackPath string) error {
+	tr := tar.NewReader(tarStream)
+
+	strip := filter.StripPath
+	target := filter.TargetPath
+
+	if target == "" {
+		op.Debugf("Bad target path in FilterSpec (%#v)", filter)
+		return fmt.Errorf("Invalid write target specified")
+	}
+
+	if strip == "" {
+		op.Debugf("Strip path was set to \"\"")
+	}
+
+	if _, err := os.Stat(unpackPath); err != nil {
+		// the target unpack path does not exist. We should not get here.
+		op.Errorf("tar unpack target does not exist (%s)", unpackPath)
+		return err
+	}
+
+	finalTargetPath := filepath.Join(unpackPath, target)
+	op.Debugf("finalized target path for Tar unpack operation at (%s)", finalTargetPath)
+
+	// process the tarball onto the filesystem
+	for {
+		header, err := tr.Next()
+		if err == io.EOF {
+			// This indicates the end of the archive
+			tarStream.Close()
```

Comment:
Yeah was thinking about this. Definitely would be safer to defer and the just break out of the loop here.